crates/bevy_ecs/src/resource.rs

```diff
@@ -72,4 +86,132 @@ pub use bevy_ecs_macros::Resource;
     label = "invalid `Resource`",
     note = "consider annotating `{Self}` with `#[derive(Resource)]`"
 )]
-pub trait Resource: Send + Sync + 'static {}
+pub trait Resource: Component<Mutability = Mutable> {}
+
+/// The `on_add` component hook that maintains the uniqueness property of a resource.
+pub fn resource_on_add_hook(mut world: DeferredWorld, context: HookContext) {
+    if let Some(&original_entity) = world.resource_entities.get(context.component_id) {
+        let name = world
```

Comment:
This does the unnecessary work of looking up the name in non-failure cases to improve the diagnostics in the failure case. This should be refactored to only do the lookup when the assertion is false (just write the check yourself and panic)